src/ui/TemplateMessageItemBody/index.tsx

```diff
@@ -0,0 +1,73 @@
+import './index.scss';
+import React, { ReactElement } from 'react';
+import type { BaseMessage } from '@sendbird/chat/message';
+import { getClassName } from '../../utils';
+import MessageTemplateProvider from '../../modules/GroupChannel/components/MessageTemplateProvider';
+import { MessageTemplateData, MessageTemplateItem, SendbirdMessageTemplate } from './types';
+import restoreNumbersFromMessageTemplateObject from './utils/restoreNumbersFromMessageTemplateObject';
+import mapData from './utils/mapData';
+import selectColorVariablesByTheme from './utils/selectColorVariablesByTheme';
+import { SendbirdTheme } from '../../types';
+
+interface Props {
+  className?: string | Array<string>;
+  message: BaseMessage;
+  isByMe?: boolean;
+  mouseHover?: boolean;
+  isReactionEnabled?: boolean;
+  theme?: SendbirdTheme;
+}
+
+/**
+ * Returns copied message template object filled with given template data and color variables.
+ */
+const getFilledMessageTemplateWithData = (
+  template: MessageTemplateItem[],
+  templateData: Record<string, any>,
+  colorVariables: Record<string, string>,
+  theme: SendbirdTheme,
+): MessageTemplateItem[] => {
+  const selectedThemeColorVariables = selectColorVariablesByTheme({
+    colorVariables,
+    theme,
+  });
+  const parsedTemplate: MessageTemplateItem[] = mapData({
+    template: restoreNumbersFromMessageTemplateObject(template) as any,
+    source: { ...templateData, ...selectedThemeColorVariables },
+  }) as any;
+  return parsedTemplate;
+};
+
+export default function TemplateMessageItemBody({
+  className = '',
+  message,
+  isByMe = false,
+  mouseHover = false,
+  isReactionEnabled = false,
+  theme = 'light',
+}: Props): ReactElement {
+  // FIXME: Is there more efficient way than below? Below parses the whole large messsage templates.
+  const allMessageTemplates: SendbirdMessageTemplate[] = JSON.parse(localStorage.getItem('message_templates')); // sb.getAllMessageTemplates();
+  if (!allMessageTemplates) return;
+
+  const templateData: MessageTemplateData = message.extendedMessagePayload?.['template'] as MessageTemplateData;
+  const template: SendbirdMessageTemplate = allMessageTemplates[templateData.key];
+  const filledMessageTemplateItems: MessageTemplateItem[] = getFilledMessageTemplateWithData(
+    template.ui_template.body.items,
+    templateData.variables,
+    template.color_variables,
+    theme,
+  );
```

Comment:
How about evaluating which one is better performance-wise and applying the superior method?

1. Traversing objects recursively to fill variables.
2. Stringifying the object using **JSON.stringify**, filling variables for the entire string, then parsing it using **JSON.parse**.

Reply:
Wow thats great idea. Lets think about it:
- Tree traversal is O(N) where N is number of nodes => option 1 is O(N)
- option 2 requires regex search for the whole string. For every regex search, match with each variable. Its O(V * regex search) Afterwards, still need to JSON.parse(str)

I think option 1 is better performance. What do you think?

Reply:
edit:
option 1 runtime: O(N)
option 2 runtime: O(regex)
because variables are flattened map. find is O(1)

they are similar but i think regex is more expensive

Reply:
It might vary depending on the structure of the template.
Let's measure the performance with actual samples.

Reply:
https://github.com/sb-repro/parse-performance/blob/main/index.ts

![image](https://github.com/sendbird/sendbird-uikit-react/assets/26326015/5f5727f5-4cb4-431f-9a39-37316d980c43)

Reply:
It has been confirmed that other platforms are also using replace+replacer similarly.
It's also the best option in terms of performance. It should be fine to use this.

https://github.com/sb-repro/parse-performance/blob/main/index.ts#L42-L60

Reply:
![Screenshot 2024-02-15 at 6 07 48 PM](https://github.com/sendbird/sendbird-uikit-react/assets/16806397/e2e290d3-9161-49e5-902c-cb5d91ff7722)
amazing! when tested locally replaceAll seems like the best option. ill apply this. thanks again!!

Reply:
If you use the regular expression `/{([^"{}]+)}/g`, the replacer will be faster! **replaceAll** is a newer syntax, so there might be some risks such as limited browser support and lack of support in React Native.

Reply:
oh thats good point. Ill take your suggestion of using regex search and replace. Ill do this after merging https://github.com/sendbird/sendbird-uikit-react/pull/979
because of dependency logic

Reply:
actually ill apply this to the child PR

Reply:
Its been applied in [this PR](https://github.com/sendbird/sendbird-uikit-react/pull/982)
but It has some styling bug for image and box item so the usage has been commented out. So it uses mapData way instead.